common/raw_hashtable_metadata_group_benchmark.cpp

```diff
@@ -0,0 +1,233 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+
+#include <algorithm>
+
+#include "absl/random/random.h"
+#include "common/raw_hashtable_metadata_group.h"
+
+namespace Carbon::RawHashtable {
+
+// If we have any SIMD support, create dedicated benchmark utilities for the
+// portable and SIMD implementation so we can directly benchmark both.
+#if CARBON_NEON_SIMD_SUPPORT || CARBON_X86_SIMD_SUPPORT
+// Override the core API with explicit use of the portable API.
+class BenchmarkPortableMetadataGroup : public MetadataGroup {
+ public:
+  explicit BenchmarkPortableMetadataGroup(MetadataGroup g) : MetadataGroup(g) {}
+
+  static auto Load(uint8_t* metadata, ssize_t index)
+      -> BenchmarkPortableMetadataGroup {
+    return BenchmarkPortableMetadataGroup(PortableLoad(metadata, index));
+  }
+  auto Store(uint8_t* metadata, ssize_t index) const -> void {
+    PortableStore(metadata, index);
+  }
+
+  auto ClearDeleted() -> void { PortableClearDeleted(); }
+
+  auto Match(uint8_t present_byte) const -> MatchRange {
+    return PortableMatch(present_byte);
+  }
+  auto MatchPresent() const -> MatchRange { return PortableMatchPresent(); }
+
+  auto MatchEmpty() const -> MatchIndex { return PortableMatchEmpty(); }
+  auto MatchDeleted() const -> MatchIndex { return PortableMatchDeleted(); }
+};
+
+// Override the core API with explicit use of the SIMD API.
+class BenchmarkSIMDMetadataGroup : public MetadataGroup {
+ public:
+  explicit BenchmarkSIMDMetadataGroup(MetadataGroup g) : MetadataGroup(g) {}
+
+  static auto Load(uint8_t* metadata, ssize_t index)
+      -> BenchmarkSIMDMetadataGroup {
+    return BenchmarkSIMDMetadataGroup(SIMDLoad(metadata, index));
+  }
+  auto Store(uint8_t* metadata, ssize_t index) const -> void {
+    SIMDStore(metadata, index);
+  }
+
+  auto ClearDeleted() -> void { SIMDClearDeleted(); }
+
+  auto Match(uint8_t present_byte) const -> MatchRange {
+    return SIMDMatch(present_byte);
+  }
+  auto MatchPresent() const -> MatchRange { return SIMDMatchPresent(); }
+
+  auto MatchEmpty() const -> MatchIndex { return SIMDMatchEmpty(); }
+  auto MatchDeleted() const -> MatchIndex { return SIMDMatchDeleted(); }
+};
+#endif
+
+namespace {
+
+constexpr ssize_t BenchSize = 256;
+
+#if CARBON_NEON_SIMD_SUPPORT || CARBON_X86_SIMD_SUPPORT
+using PortableGroup = BenchmarkPortableMetadataGroup;
+using SIMDGroup = BenchmarkSIMDMetadataGroup;
+#endif
+
+struct BenchMetadata {
+  llvm::MutableArrayRef<uint8_t> metadata;
+
+  // For random byte metadata, also store the specific byte to search for in
+  // each group here. For other kinds, this is empty.
+  llvm::ArrayRef<uint8_t> bytes;
```

Comment:
Is this an optional byte? or is there a case when there is more than one? Looking below, it has `BenchSize` bytes when non-empty.

Please clarify the comment.